# DS-SF-34 | 02 | The `pandas` Library | Assignment | Starter Code

## The Bistro Meets `pandas`

You've just told one of your friend that you are taking a Data Science class.  (Yeah!)  Your friend is running a bistro, a small restaurant, serving moderately priced simple meals in a modest setting ([Wikipedia](https://en.wikipedia.org/wiki/Bistro)).  She collected over some period of time the following information of her patrons' visits.

| Variable's name | Its meaning |
|:---:|:---|
| `name` | Patron's first name |
| `gender` | Patron's gender |
| `is_smoker` | Whether the patron is smoking or not |
| `party` | Party's size |
| `check` | Check amount (\$) (after taxes but before tips) |
| `tip` | Tip (\$) that the patron added to the check |
| `day` | Week day of the visit |
| `time` | Rough time estimate of the visit |

In this assignment, we will be exploring this dataset using `pandas`.<sup>(*)</sup>

<sup>(*)</sup> this dataset was adapted from the `tips` dataset of the `seaborn` package (https://github.com/mwaskom/seaborn-data)

> ### Question 1.  Import `numpy` (as `np`) and `pandas` (as `pd`).

In [21]:
import os

# TODO
import pandas as pd
import numpy as np

pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

> ### Question 2.  Read the `dataset-02-tips.csv` dataset.

In [22]:
# TODO
pd.read_csv('../datasets/dataset-02-bistro.csv')

day    time      name  gender is_smoker  party  check   tip
0      Sunday  Dinner  Kimberly  Female     False      2  16.99  1.01
1      Sunday  Dinner  Nicholas    Male     False      3  10.34  1.66
2      Sunday  Dinner     Larry    Male     False      3  21.01  3.50
3      Sunday  Dinner    Joseph    Male     False      2  23.68  3.31
4      Sunday  Dinner    Janice  Female     False      4  24.59  3.61
..        ...     ...       ...     ...       ...    ...    ...   ...
239  Saturday  Dinner     Kevin    Male     False      3  29.03  5.92
240  Saturday  Dinner    Sandra  Female     False      2  27.18  2.00
241  Saturday  Dinner      Carl    Male     False      2  22.67  2.00
242  Saturday  Dinner       Jon    Male     False      2  17.82  1.75
243  Thursday  Dinner    Brandi  Female     False      2  18.78  3.00

[244 rows x 8 columns]

> ### Question 3.  What is the class of the `pandas` object storing the dataset?

In [23]:
# TODO
type(df)

NameError: name 'df' is not defined

Answer: TODO

> ### Question 4.  How many samples (i.e., rows) are in this dataset?

In [18]:
# TODO
df.shape

(244, 7)

Answer: 244 rows

> ### Question 5.  How many variables (i.e., columns) are in this dataset?

In [20]:
# TODO
df.shape

NameError: name 'df' is not defined

Answer: 7

> ### Question 6.  Print the name of each column in the dataset, one name per line.

In [20]:
# TODO
df.columns

Index([u'total_bill', u'tip', u'sex', u'smoker', u'day', u'time', u'size'], dtype='object')

> ### Question 7.  Print the first two rows of the dataset to the console.  What does the output look like?

In [22]:
# TODO
df[:2]

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3

Answer: TODO

> ### Question 8.  Extract the last 2 rows of the data frame and print them to the console.  What does the output look like?

In [23]:
# TODO
df[-2:]

total_bill   tip     sex smoker   day    time  size
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

Answer: TODO

> ### Question 9.  Does the dataset contain any missing values?

In [28]:
# TODO
df.isnull().sum().sum()

0

Answer: No

> ### Question 10.  What can you say about the `is_smoker` variable?  I.e., will it bring any insights when analyzing the dataset?  What do you want to do with it?  (and do it...)

In [32]:
df.columns

Index([u'total_bill', u'tip', u'sex', u'smoker', u'day', u'time', u'size'], dtype='object')

In [9]:
# TODO

Answer: You can see how many of your customers are somkers.

> ### Question 11.  For which week days does the dataset has data for?

In [ ]:
# TODO

Answer: TODO

> ### Question 12.  How often was the bistro patronized for each week day?

(check `.value_counts()`; it could come in handy)

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html)

In [ ]:
# TODO

Answer: TODO

> ### Question 13.  How much tip did waiters collect for each week day?

In [ ]:
# TODO

Answer: TODO

> ### Question 14.  What is the average tip per check (in absolute \$) for each week day?

In [ ]:
# TODO

Answer: TODO

> ### Question 15.  What is the average tip per check (as a percentage of the check) for each week day?

In [ ]:
# TODO

Answer: TODO

> ### Question 16.  Are there any name in common between male and female patrons?  (E.g., `Chris` can refer to either a man or a woman)

(check `numpy.intersect1d()`; it could come in handy)

(https://docs.scipy.org/doc/numpy/reference/generated/numpy.intersect1d.html)

In [ ]:
# TODO

Answer: TODO

> ### Question 17.  If no patrons share the same name, how many unique patrons are in the dataset?

In [ ]:
# TODO

Answer: TODO

> ### Question 18.  How many times did `Kevin` patronized the bistro?  How about `Alice`?

In [ ]:
# TODO

Answer: TODO

> ### Question 19.  Who are the top 3 female and male patrons?

In [ ]:
# TODO

Answer: TODO

> ### Question 20.  Who's the best tipper (as a fraction of all tips over all check totals)?  Who's the worst?  How many times did they patronize the bistro?

(check `numpy.intersect1d()`; it could come in handy)

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.sort_values.html)

In [ ]:
# TODO

Answer: TODO